jerry-ext/args/args.c

```diff
@@ -0,0 +1,97 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+/**
+ * It validate the JS arguments and assign them to the native argument.
+ *
+ * @return jerry undefined: the validator passes,
+ *         jerry error: the validator fails.
+ */
+jerry_value_t
+jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p, /**< points to the js argument */
+                                 const jerry_length_t js_arg_cnt, /**< the count of the js arguments */
+                                 const jerryx_arg_t *c_arg_p, /**< points to the jerryx_arg_t */
+                                 const jerry_length_t c_arg_cnt) /**< the count of the native argument */
+{
+  jerry_value_t ret = jerry_create_undefined ();
+
+  if (c_arg_cnt == 0)
+  {
+    return ret;
+  }
+
+  jerryx_arg_iterator_t iterator =
```

Comment:
The `jerryx_arg_iterator_t` struct does not feel right to me.
Passing that struct to the transforms exposes too much of the internals of the validate-and-assign algorithm.
All they should care about what the JS argument is and where to store the C value (if possible).
What the internal pointers and counters are should not be of their concern.

I think it is possible to perform all the here-shown transforms properly even without the iterator.
Therefore, I'd suggest dropping `jerryx_arg_iterator_t` and changing the `jerryx_arg_transform_func_t`signature.
(Algorithm variant also proposed below.)

```c
typedef jerry_value_t (*jerryx_arg_transform_func_t) (const jerryx_arg_t *c_arg_p, jerry_value_t js_arg);

jerry_value_t
jerryx_arg_transform_args (const jerry_value_t *js_arg_p,
                           jerry_length_t js_arg_cnt,
                           const jerryx_arg_t *c_arg_p,
                           jerry_length_t c_arg_cnt)
{
  jerry_value_t ret = jerry_create_undefined ();
  
  for (; c_arg_cnt != 0 && !jerry_value_has_error_flag (ret); c_arg_cnt--, c_arg_p++, js_arg_cnt--, js_arg_p++)
  {
    ret = c_arg_p->func (c_arg_p, js_arg_cnt > 0 ? *js_arg_p : jerry_create_undefined ());
  }

  return ret;
}
```

This could also make things easier:
- could get rid of helpers: `jerryx_arg_iterator_pop`, `jerryx_arg_iterator_peek`, even `jerryx_arg_iterator_get_dest` perhaps.
- no need for all transforms to start with `jerry_value_t js_arg = jerryx_arg_iterator_pop (iterator_p);`

Let me know what you think. (E.g., if I missed some corner cases.)